apps/prairielearn/src/ee/auth/lti13/lti13Auth.ts

```diff
@@ -0,0 +1,254 @@
+import { Router, type Request, Response, NextFunction } from 'express';
+import asyncHandler = require('express-async-handler');
+import { Issuer, Strategy, type StrategyVerifyCallbackReq, IdTokenClaims } from 'openid-client';
+import * as passport from 'passport';
+import { z } from 'zod';
+import { get as _get } from 'lodash';
+
+import { loadSqlEquiv, queryAsync } from '@prairielearn/postgres';
+import * as error from '@prairielearn/error';
+import * as authnLib from '../../../lib/authn';
+import { selectLti13Instance } from '../../models/lti13Instance';
+import { get as cacheGet, set as cacheSet } from '../../../lib/cache';
+import { getInstitutionAuthenticationProviders } from '../../lib/institution';
+
+const sql = loadSqlEquiv(__filename);
+const router = Router({ mergeParams: true });
+
+// Middleware to check access
+router.use(
+  asyncHandler(async (req, res, next) => {
+    const lti13_instance = await selectLti13Instance(req.params.lti13_instance_id);
+    const instAuthProviders = await getInstitutionAuthenticationProviders(
+      lti13_instance.institution_id,
+    );
+
+    if (
+      !instAuthProviders.some((a) => {
+        return a.name === 'LTI 1.3';
+      })
+    ) {
+      throw error.make(404, 'Institution does not support LTI 1.3 authentication');
+    }
+
+    next();
+  }),
+);
+
+// Express routes
+router.get('/login', launchFlow);
+router.post('/login', launchFlow);
+router.post(
+  '/callback',
+  asyncHandler(async (req, res) => {
+    const lti13_instance = await selectLti13Instance(req.params.lti13_instance_id);
+
+    req.session.lti13_claims = await authenticate(req, res);
```

Comment:
Correct, storing the lti13_claims in the session is not used for authentication (this PR) but later ones use some of that authorization data to make choices. I can refactor this code so that it's not by default storing to req.session and we can one-liner add that later when it's used.

Reply:
Sounds good, I'd appreciate that for now. Start thinking about the story with multiple LTI instances!

Reply:
Addressed in 593dd5

Reply:
I don't think multiple LTI instances is a challenge. Putting it in the session helps more for the authorization flow but it might not even be necessary after a certain point (like once the student is in the course_instance). Or at least that's my design hope -- I don't see the lti13_claims as being a long running persistent thing.

If that authorization does something, like enroll them in a course, it's probably a one-and-done kind of thing. I don't think multiple concurrent logins is a challenge - the user would just go to their other course in the LMS, follow the links there, and it would do whatever is needed.